multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakSettings.java

```diff
@@ -13,48 +13,131 @@
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.collect.Maps;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.Map;
 import javax.inject.Inject;
 import javax.inject.Named;
 import javax.inject.Singleton;
 import org.eclipse.che.commons.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** @author Max Shaposhnik (mshaposh@redhat.com) */
 @Singleton
 public class KeycloakSettings {
+  private static final Logger LOG = LoggerFactory.getLogger(KeycloakSettings.class);
 
   private final Map<String, String> settings;
 
   @Inject
   public KeycloakSettings(
-      @Named(AUTH_SERVER_URL_SETTING) String serverURL,
-      @Named(REALM_SETTING) String realm,
+      @Nullable @Named(JS_ADAPTER_URL_SETTING) String jsAdapterUrl,
+      @Nullable @Named(AUTH_SERVER_URL_SETTING) String serverURL,
+      @Nullable @Named(REALM_SETTING) String realm,
       @Named(CLIENT_ID_SETTING) String clientId,
+      @Nullable @Named(OIDC_PROVIDER_SETTING) String oidcProvider,
+      @Named(USE_NONCE_SETTING) boolean useNonce,
       @Nullable @Named(OSO_ENDPOINT_SETTING) String osoEndpoint,
       @Nullable @Named(GITHUB_ENDPOINT_SETTING) String gitHubEndpoint) {
+
+    if (serverURL == null && oidcProvider == null) {
+        throw new RuntimeException("Either the '" + AUTH_SERVER_URL_SETTING + "' or '" + OIDC_PROVIDER_SETTING + "' property should be set");
+    }
+    
+    if (oidcProvider == null && realm == null) {
+        throw new RuntimeException("The '" + REALM_SETTING + "' property should be set");
+    }
+    
+    String wellKnownEndpoint = oidcProvider != null ? oidcProvider : serverURL + "/realms/" + realm;
+    if (!wellKnownEndpoint.endsWith("/")) {
+      wellKnownEndpoint = wellKnownEndpoint + "/";
+    }
+    wellKnownEndpoint += ".well-known/openid-configuration";
+
+    LOG.info("Retrieving OpenId configuration from endpoint: {}", wellKnownEndpoint);
+
+    URL url;
+    Map<String, Object> openIdConfiguration;
+    try {
+      url = new URL(wellKnownEndpoint);
+      final InputStream inputStream = url.openStream();
+      final JsonFactory factory = new JsonFactory();
+      final JsonParser parser = factory.createParser(inputStream);
+      final TypeReference<Map<String, Object>> typeReference =
+          new TypeReference<Map<String, Object>>() {};
+      openIdConfiguration = new ObjectMapper().reader().readValue(parser, typeReference);
+    } catch (IOException e) {
+      LOG.error(
+          "Exception while retrieving OpenId configuration from endpoint: " + wellKnownEndpoint, e);
+      this.settings = Collections.unmodifiableMap(new HashMap<>());
+      return;
+    }
+
+    LOG.info("openid configuration = {}", openIdConfiguration);
```

Comment:
LOG.DEBUG?

Reply:
Well, I didn't change this since your first approval (https://github.com/eclipse/che/pull/8650#pullrequestreview-94662074).

I can change this to `DEBUG` level if you really want, but since it is only shown at startup it doesn't hurt displaying in the `INFO` level IMO.

OTOH, I think it might be interesting to have these authentication details available in order to be able to troubleshoot future authentication problems without the need to restart the Che server (which we know is not easy currently since it requires stopping all workspaces of all users)

@mario any thought on this ?

Reply:
up to you and @l0rd

Reply:
@davidfestal @skabashnyuk ok to leave like this for me

Reply:
@skabashnyuk are you OK to approve now ?

Reply:
what about this comment https://github.com/eclipse/che/pull/8650#discussion_r176007211?

Reply:
@skabashnyuk 

Sorry, I missed this one.Fixed in commit https://github.com/eclipse/che/pull/8650/commits/b6a52430821d8c29e158a53f5a848dd0a5291143